File: ow/dos_.pas
(* dos_.pas -- Pascal declarations for dos_.c.

   This file is for linking compiled object files with Pascal linker.
   It will be deleted in future when we rewrite the project in C.

   This is free and unencumbered software released into the public domain.
   For more information, please refer to <http://unlicense.org>. *)

unit dos_;

interface

function _dos_allocmem(size: word; var seg: word): word;
function _dos_freemem(seg: word): word;
function _dos_setblock(size: word; seg: word; var max: word): word;

implementation

uses
    pascal,
    i86;

(*$l dos_.obj*)
function _dos_allocmem(size: word; var seg: word): word; external;
function _dos_freemem(seg: word): word; external;
function _dos_setblock(size: word; seg: word; var max: word): word; external;

end.

File: src/ow/dos_.c
/* dos_.c -- custom DOS library.

   This is free and unencumbered software released into the public domain.
   For more information, please refer to <http://unlicense.org>. */

#ifdef __WATCOMC__
#include <i86.h>
#include <stddef.h>
#include <stdbool.h>
#include <stdint.h>
#endif

#include "pascal/pascal.h"
#include "ow/dos_.h"

uint16_t __near __doserror(uint16_t code)
{
    return code ? code : 0xffff;
}

uint16_t PUBLIC_CODE _dos_allocmem(uint16_t size, uint16_t *seg)
{
    union REGPACK regs;
    regs.w.bx = size;
    regs.h.ah = 0x48;
    intr(0x21, &regs);
    if (regs.w.flags & INTR_CF)
    {
        *seg = regs.w.bx;
        return __doserror(regs.w.ax);
    }
    else
    {
        *seg = regs.w.ax;
        return 0;
    };
}

uint16_t PUBLIC_CODE _dos_freemem(uint16_t seg)
{
    union REGPACK regs;
    regs.w.es = seg;
    regs.h.ah = 0x49;
    intr(0x21, &regs);
    if (regs.w.flags & INTR_CF)
        return __doserror(regs.w.ax);
    else
        return 0;
}

uint16_t PUBLIC_CODE _dos_setblock(uint16_t size, uint16_t seg, uint16_t *max)
{
    union REGPACK regs;
    regs.w.bx = size;
    regs.w.es = seg;
    regs.h.ah = 0x4a;
    intr(0x21, &regs);
    if (regs.w.flags & INTR_CF)
    {
        *max = regs.w.bx;
        return __doserror(regs.w.ax);
    }
    else
        return 0;
}
